File: experiments/java-sdk/src/main/java/com/datadog/llmobs/experiments/internal/SpanBuilder.java
package com.datadog.llmobs.experiments.internal;

import com.datadog.llmobs.experiments.ExperimentResult;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Maps a row result to the JSON shape expected by the experiment events endpoint.
 *
 * <p>v0.1 builds raw {@code Map<String, Object>} structures rather than the generated
 * {@code LLMObsExperimentSpan} model. The model class types {@code expected_output} as
 * {@code Map<String, Object>}, which forces a wrapper ({@code {"value": ...}}) around any
 * non-Map value. The live API accepts the raw value (string, number, map, list) directly,
 * matching the Python SDK's behavior — so we bypass the typed model and send raw.
 */
public final class SpanBuilder {

    private SpanBuilder() {}

    public static Map<String, Object> toSpan(
        ExperimentResult.Row row,
        Map<String, Object> recordMetadata,
        String projectId,
        String datasetId,
        String datasetRecordId,
        String experimentId,
        String spanName,
        Map<String, String> userTags
    ) {
        Map<String, Object> meta = new LinkedHashMap<>();
        if (row.input() != null) {
            meta.put("input", row.input());
        }
        if (row.output() != null) {
            meta.put("output", row.output());
        }
        if (row.expectedOutput() != null) {
            meta.put("expected_output", row.expectedOutput());
        }
        if (recordMetadata != null && !recordMetadata.isEmpty()) {
            meta.put("metadata", recordMetadata);
        }
        if (row.isError()) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("type", row.errorType());
            err.put("message", row.errorMessage());
            meta.put("error", err);
        }

        Map<String, Object> span = new LinkedHashMap<>();
        span.put("span_id", row.spanId());
        span.put("trace_id", row.traceId());
        span.put("project_id", projectId);
        span.put("dataset_id", datasetId);
        span.put("name", spanName);
        span.put("start_ns", row.startNs());
        span.put("duration", row.durationNs());
        span.put("status", row.isError() ? "error" : "ok");
        span.put("meta", meta);
        span.put("tags", buildTags(userTags, experimentId, datasetId, datasetRecordId));
        return span;
    }

    private static List<String> buildTags(
        Map<String, String> userTags,
        String experimentId,
        String datasetId,
        String datasetRecordId
    ) {
        Map<String, String> merged = new LinkedHashMap<>();
        if (userTags != null) {
            merged.putAll(userTags);
        }
        merged.put("experiment_id", experimentId);
        merged.put("dataset_id", datasetId);
        if (datasetRecordId != null && !datasetRecordId.isEmpty()) {
            merged.put("dataset_record_id", datasetRecordId);
        }
        List<String> out = new ArrayList<>(merged.size());
        for (Map.Entry<String, String> e : merged.entrySet()) {
            out.add(e.getKey() + ":" + e.getValue());
        }
        return out;
    }
}

File: experiments/node-sdk/src/internal/spanBuilder.ts
import type { Row } from "../ExperimentResult.js";

/**
 * Build the raw span map for one experiment row (the LLM Obs experiment span
 * wire format).
 *
 * Tag precedence: the auto tags `experiment_id`, `dataset_id`,
 * `dataset_record_id` always win over user-supplied tags on key conflict.
 * `metadata` lives only in `meta.metadata` — never in tags.
 */
export function toSpan(
  row: Row,
  metadata: Record<string, unknown>,
  experimentId: string,
  projectId: string,
  datasetId: string,
  datasetRecordId: string,
  spanName: string,
  userTags: Record<string, string>,
): Record<string, unknown> {
  const meta: Record<string, unknown> = {
    input: row.input ?? null,
    output: row.output ?? null,
    expected_output: row.expectedOutput ?? null,
  };

  // metadata: raw / unwrapped. Only emitted when present (W4).
  if (metadata && Object.keys(metadata).length > 0) {
    meta.metadata = metadata;
  }

  if (row.isError) {
    meta.error = {
      type: row.errorType ?? "",
      message: row.errorMessage ?? "",
      stack: "",
    };
  }

  return {
    span_id: row.spanId,
    trace_id: row.traceId,
    project_id: projectId,
    dataset_id: datasetId,
    name: spanName,
    start_ns: row.startNs,
    duration: row.durationNs,
    status: row.isError ? "error" : "ok",
    meta,
    tags: buildTags(userTags, experimentId, datasetId, datasetRecordId),
  };
}

function buildTags(
  userTags: Record<string, string>,
  experimentId: string,
  datasetId: string,
  datasetRecordId: string,
): string[] {
  const tags = new Map<string, string>();
  // User tags first, so the auto tags below overwrite on conflict.
  for (const [k, v] of Object.entries(userTags ?? {})) {
    tags.set(k, `${k}:${v}`);
  }
  tags.set("experiment_id", `experiment_id:${experimentId}`);
  tags.set("dataset_id", `dataset_id:${datasetId}`);
  tags.set("dataset_record_id", `dataset_record_id:${datasetRecordId}`);
  return [...tags.values()];
}
